src/WeightVectors.jl

```diff
@@ -212,10 +212,33 @@ Base.iszero(w::AbstractWeightVector) = w.m[2] == 5
     sample_within_level(rng, m, pos, len)
 end
 
-@inline function sample_within_level(rng, m, pos, len)
-    while true
+macro inbounds_if_word64(arg)
+    esc(Sys.WORD_SIZE == 64 ? :(@inbounds $arg) : arg)
+end
+
+@inline function sample_within_level(rng, m, pos::UInt64, len::UInt64)
+    shift = leading_zeros(len-1)
+    # @inbounds safety when Sys.WORD_SIZE == 64: 
+    # 1. 0 <= r <= typemax(UInt64) since r = rand(rng, UInt64)
+    # 2. len and pos ∈ [1, 2^56]. This is explicitly checked and needs to be
+    #    true if m is not corrupted since the maximum length of m is 10794 + 8 * 2^52
+    #    and pos is the absolute position of the group into m and len is the length of 
+    #    the group.
+    # 3. shift = leading_zeros(len-1) which means that shift ∈ [8, 64].
+    # 4. In the loop, k1 = (r >> shift) and k2 = _convert(Int, k1<<1+pos) so 
+    #    k2+1 = _convert(Int, (r >> shift) << 1 + pos) + 1 where the conversion can be
+    #    spared since (r >> shift) << 1 + pos <= 2^63-1.
+    # 5. max(k2+1) = (typemax(UInt64) >> shift) << 1 + pos + 1 since max((r >> shift) << 1) = 
+    #    (max(r) >> shift) << 1 = (typemax(UInt64) >> shift) << 1 by 1.,2. and 3. 
+    #    Similarly, min(k2+1) = pos + 1 >= 2 by 1.,2. and 3.
+    # 6. Therefore, both m[k2] and m[k2+1] are inbound if we check that 2. holds and if we check
+    #    that max(k2+1) is inbounds. If one of these two conditions doesn't hold, m is corrupted and
+    #    then we throw an error.
+    (UInt64(1) <= len <= UInt64(2)^56 && UInt64(1) <= pos <= UInt64(2)^56) || throw("Sampler is corrupted")
+    checkbounds(m, (typemax(UInt64) >> shift) << 1 + pos + 1)
+    @inbounds_if_word64 while true
         r = rand(rng, UInt64)
-        k1 = (r>>leading_zeros(len-1))
+        k1 = (r>>shift)
         k2 = _convert(Int, k1<<1+pos)
         # TODO for perf: delete the k1 < len check by maintaining all the out of bounds m[k2] equal to 0
         rand(rng, UInt64) < m[k2] * (k1 < len) && return Int(signed(m[k2+1]))
```

Comment:
```suggestion
        rand(rng, UInt64) < @inbounds(m[k2]) * (k1 < len) && return Int(signed(@inbounds(m[k2+1])))
```
It's best to keep the smallest scope possible for `@inbounds`. This could matter if, for example, someone defines a custom rng that has a `rand(rng, ::Type{UInt64})` method that does necessary bounds checking on `rng` which can be skipped with `@inbounds`.